File: 2-Graph Search, Shortest Path, and Data Structures/3-Graph Search/Java/BFSApp.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

import directed_graph.DirectedGraph;
import graph.GraphInterface;
import undirected_graph.UndirectedGraph;

/**
 * BFS Applications.
 * 1. Naive BFS
 * 2. Shortest path (specific to BFS)
 * 3. Undirected connectivity
 *
 * @author Ziang Lu
 */
public class BFSApp {

    /**
     * Main driver.
     * @param args arguments from command line
     */
    public static void main(String[] args) {
        System.out.println("Testing undirected graph...");
        testUndirectedGraph("undirected_graph_info.txt");
        System.out.println();

        System.out.println("Testing directed graph...");
        testDirectedGraph("directed_graph_info.txt");
    }

    /**
     * Test driver for BFS with undirected graph.
     * @param filename undirected graph filename
     */
    private static void testUndirectedGraph(String filename) {
        // Construct the graph
        GraphInterface graph = constructGraph(filename, true);

        // Find all the findable vertices starting from vertex #1 using BFS
        System.out.println("Findable vertices from vertex #1 using BFS: " + graph.bfs(1)); // [1, 2, 3, 4, 5, 6]
        graph.clearExplored();

        // Find the length of the shortest path from vertex #1 to vertex #6
        System.out.println("Length of the shortest path from vertex #1 to vertex #6: " + graph.shortestPath(1, 6)); // 3
        graph.clearExplored();

        // Find the number of connected components of the undirected graph using BFS
        System.out.println(
                "Number of connected components of the graph using BFS: " + graph.numOfConnectedComponentsWithBFS()); // 1
    }

    /**
     * Test driver for BFS with directed graph.
     * @param filename directed graph filename
     */
    private static void testDirectedGraph(String filename) {
        // Construct the graph
        GraphInterface graph = constructGraph(filename, false);

        // Find all the findable vertices starting from vertex #1 using BFS
        System.out.println("Findable vertices from vertex #1 using BFS: " + graph.bfs(1)); // [1, 2, 3]
        graph.clearExplored();

        // Find the length of the shortest path from vertex #1 to vertex #3
        System.out.println("Length of the shortest path from vertex #1 to vertex #3: " + graph.shortestPath(1, 3)); // 1
        graph.clearExplored();
    }

    /**s
     * Private helper method to construct a undirected graph from the given file.
     * @param filename given filename
     * @param undirected whether the graph to construct is undirected
     * @return constructed undirected graph
     */
    private static GraphInterface constructGraph(String filename, boolean undirected) {
        GraphInterface graph = null;

        Scanner scanner = null;
        try {
            scanner = new Scanner(new File(filename), "latin1");

            if (undirected) {
                graph = new UndirectedGraph();
            } else {
                graph = new DirectedGraph();
            }
            // Add the vertices
            int nVtx = Integer.parseInt(scanner.nextLine());
            for (int vtxID = 1; vtxID <= nVtx; ++vtxID) {
                graph.addVtx(vtxID);
            }
            // Add the edges
            while (scanner.hasNextLine()) {
                String[] ends = scanner.nextLine().split(" ");
                graph.addEdge(Integer.parseInt(ends[0]), Integer.parseInt(ends[1]));
            }
        } catch (FileNotFoundException ex) {
            System.out.println("Cannot found the file.");
        } finally {
            if (scanner != null) {
                scanner.close();
            }
        }

        return graph;
    }

}

File: 2-Graph Search, Shortest Path, and Data Structures/3-Graph Search/Python/bfs_app.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

__author__ = 'Ziang Lu'

from directed_graph import DirectedGraph
from undirected_graph import UndirectedGraph


def main():
    """
    Main driver.
    :return: None
    """
    _test_undirected_graph(filename='undirected_graph_info.txt')


def _test_undirected_graph(filename):
    """
    Test driver for undirected graph with BFS.
    :param filename: str
    :return: None
    """
    # Construct the graph
    graph = _construct_graph(filename=filename, undirected=True)

    # Find all the findable vertices starting from vertex #1 using BFS
    print('Findable vertices from vertex #1 using BFS: %s' %
          graph.bfs(src_vtx_id=1))  # [1, 2, 3, 4, 5, 6]
    graph.clear_explored()

    # Find the length of the shortest path from vertex #1 to vertex #6
    print('Length of the shortest path from vertex #1 to vertex #6: %d' %
          graph.shortest_path(src_vtx_id=1, dest_vtx_id=6))  # 3
    graph.clear_explored()

    # Find the number of connected components of the undirected graph using BFS
    print('Number of connected components using BFS: %d' %
          graph.num_of_connected_components_with_bfs())  # 1


def _test_directed_graph(filename):
    """
    Test driver for directed graph with BFS.
    :param filename: str
    :return: None
    """
    # Construct the graph
    graph = _construct_graph(filename=filename, undirected=False)

    # Find all the findable vertices from vertex #1 using BFS
    print('Findable vertices from vertex #1 using BFS: %s' %
          graph.bfs(src_vtx_id=1))  # [1, 2, 3]
    graph.clear_explored()

    # Find the length of the shortest path from vertex #1 to vertex #6
    print('Length of the shortest path from vertex #1 to vertex #3: %d' %
          graph.shortest_path(src_vtx_id=1, dest_vtx_id=3))  # 1


def _construct_graph(filename, undirected):
    """
    Private helper method to construct a graph from the given graph file.
    :param filename: str
    :param undirected: bool
    :return: UndirectedGraph or DirectedGraph
    """
    with open(filename, 'rt') as f:
        # Construct the graph
        if undirected:
            graph = UndirectedGraph()
        else:
            graph = DirectedGraph()
        # Add the vertices
        n_vtx = int(f.readline())
        for vtx_id in range(1, n_vtx + 1):
            graph.add_vtx(new_vtx_id=vtx_id)
        # Add the edges
        for line in f.readlines():
            ends = line.split(' ')
            graph.add_edge(end1_id=int(ends[0]), end2_id=int(ends[1]))
        return graph


if __name__ == '__main__':
    main()
